File: frontend/src/types/health.ts
// Mirrors backend/src/api/models.py's Health section field-for-field.
//
// `counts` is a plain dict on the backend (`counts: dict`), keyed by Neo4j node label plus
// two extras — verified live against GET /api/health:
//   Person, Decision, Message, Organization, Evidence, Claim, Deal, total_edges, vectors
// Typed as an index signature rather than a fixed shape because the backend builds it
// dynamically from whatever labels exist in the graph.
//
// Day 42: `report` was added to the backend response (see the diff to
// backend/src/api/models.py — `HealthResponse.report: Optional[dict]`), carrying the full
// HealthMonitor.full_health_report() output used by the dashboard's cards and charts. It can
// be `null` if that report failed to generate on the backend even though the lightweight
// connectivity check (status/services/counts, unchanged) succeeded — every reader of
// `report.*` must null-check first.

export interface ServiceStatus {
  name: string
  status: string // "ok" | "error"
  detail: string | null
}

// ---------------------------------------------------------------------------------------
// report.graph_size
// ---------------------------------------------------------------------------------------

export interface HealthGraphSize {
  nodes: Record<string, number> // Person, Decision, Message, Organization, Evidence, Claim, Deal, TOTAL
  edges: Record<string, number> // SENT_TO, AFFECTS, MADE_BY, ..., TOTAL
}

// ---------------------------------------------------------------------------------------
// report.claim_quality
// ---------------------------------------------------------------------------------------

export interface EvidenceCoverage {
  no_evidence: number
  single_evidence: number
  multi_evidence: number
  avg_evidence_per_claim: number
}

export interface EvidenceVerification {
  total: number
  verified: number
  unverified: number
  unknown: number
  verification_rate: number // already a percentage, e.g. 96.7
}

export interface ClaimQuality {
  average_confidence: number // 0-1
  min_confidence: number
  max_confidence: number
  total_claims: number
  // Keys are confidence-bucket labels ("0.90-1.00", "0.70-0.89", ...). On the current corpus
  // only 2 of the possible 5 buckets are present at all — a bucket with zero claims is simply
  // absent as a key, not present with value 0. Never assume all 5 exist.
  confidence_distribution: Record<string, number>
  claims_by_type: Record<string, number> // works_with, requests_from, informs, reports_to, negotiating_with
  evidence_coverage: EvidenceCoverage
  evidence_verification: EvidenceVerification
}

// ---------------------------------------------------------------------------------------
// report.temporal_health
// ---------------------------------------------------------------------------------------

export interface TemporalHealth {
  claims_by_status: Record<string, number> // current, review, superseded
  supersession_edges: number
  conflict_pairs: number
  claims_with_closed_windows: number
  undated_claims: number
  date_range: { earliest: string; latest: string }
}

// ---------------------------------------------------------------------------------------
// report.access_levels
// ---------------------------------------------------------------------------------------

export type AccessLevelBreakdown = Record<string, number> // PUBLIC, INTERNAL, CONFIDENTIAL, RESTRICTED

export type AccessLevels = Record<string, AccessLevelBreakdown> // keyed by node label (Claim, Evidence, ...)

// ---------------------------------------------------------------------------------------
// report.entity_stats
// ---------------------------------------------------------------------------------------

export interface PersonStats {
  total: number
  avg_mentions: number
  max_mentions: number
  total_aliases: number
  total_emails: number
}

export interface OrganizationStats {
  total: number
  avg_mentions: number
}

export interface TopPerson {
  name: string
  mentions: number
}

export interface EntityStats {
  persons: PersonStats
  organizations: OrganizationStats
  org_types: Record<string, number>
  top_persons: TopPerson[]
}

// ---------------------------------------------------------------------------------------
// report.data_quality
// ---------------------------------------------------------------------------------------

export interface DataQuality {
  claims_without_subject: number
  claims_without_object: number
  claims_without_evidence: number
  evidence_without_message: number
  messages_without_sender: number
  decisions_without_maker: number
  decisions_with_unresolved_affects: number
  total_unresolved_affects_strings: number
  soft_deleted_nodes: number
  quality_score: number // 0-100
}

// ---------------------------------------------------------------------------------------

export interface HealthReport {
  timestamp: string
  graph_size: HealthGraphSize
  claim_quality: ClaimQuality
  temporal_health: TemporalHealth
  access_levels: AccessLevels
  entity_stats: EntityStats
  data_quality: DataQuality
}

export interface HealthResponse {
  status: string // "healthy" | "degraded"
  services: ServiceStatus[]
  counts: Record<string, number>
  report: HealthReport | null
}

// ---------------------------------------------------------------------------------------
// Admin — mirrors backend/src/api/models.py's Admin section. Only the two counts (`total`)
// are used by the Day 42 health dashboard today; the full item lists are here too since
// they're 1:1 with the backend response and will be what Days 43-44's conflict/review pages
// build on, rather than being redefined there.
// ---------------------------------------------------------------------------------------

export interface ConflictItem {
  claim_id: string
  claim_type: string
  subject_name: string
  object_name: string
  conflicts_with: string[]
  confidence: number
}

export interface ConflictListResponse {
  conflicts: ConflictItem[]
  total: number
}

export interface ReviewItem {
  claim_id: string
  claim_type: string
  subject_name: string
  object_name: string
  status: string
  confidence: number
  reason: string
}

export interface ReviewQueueResponse {
  items: ReviewItem[]
  total: number
}

File: backend/src/api/routes/health.py
"""
Health route — system status dashboard endpoint.

GET /api/health — checks Neo4j and Qdrant connectivity,
                  returns node/edge/vector counts.

Powers the health dashboard (Day 42) and is useful for
monitoring and debugging.
"""

import logging

from fastapi import APIRouter, Depends
from src.graph.health_monitor import HealthMonitor
from src.api.dependencies import get_neo4j_driver, get_qdrant_index
from src.api.models import HealthResponse, ServiceStatus
from src.retrieval.qdrant_index import QdrantIndex

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("/health", response_model=HealthResponse)
async def health_check(
    driver=Depends(get_neo4j_driver),
    qdrant: QdrantIndex = Depends(get_qdrant_index),
):
    """
    Check the health of all backend services.

    Returns:
      - Status of Neo4j and Qdrant connections
      - Node counts by label
      - Edge count
      - Vector count in Qdrant
    """
    services = []
    counts = {}
    overall_healthy = True

    # Check Neo4j
    try:
        with driver.session() as session:
            # Connectivity check
            session.run("RETURN 1").single()

            # Node counts by label
            result = session.run("""
                MATCH (n)
                WHERE n.is_deleted = false
                WITH labels(n)[0] AS label, count(n) AS cnt
                RETURN label, cnt
                ORDER BY cnt DESC
            """)
            for r in result:
                counts[r["label"]] = r["cnt"]

            # Edge count
            edge_result = session.run("""
                MATCH ()-[r]->()
                RETURN count(r) AS total_edges
            """)
            counts["total_edges"] = edge_result.single()["total_edges"]

        services.append(ServiceStatus(
            name="neo4j", status="ok",
            detail=f"{sum(v for k, v in counts.items() if k != 'total_edges')} nodes, "
                   f"{counts.get('total_edges', 0)} edges",
        ))

    except Exception as e:
        logger.error("Neo4j health check failed: %s", e)
        services.append(ServiceStatus(
            name="neo4j", status="error", detail=str(e)
        ))
        overall_healthy = False

    # Check Qdrant
    try:
        info = qdrant.get_collection_info()
        counts["vectors"] = info["points_count"]

        services.append(ServiceStatus(
            name="qdrant", status="ok",
            detail=f"{info['points_count']} vectors, status={info['status']}",
        ))

    except Exception as e:
        logger.error("Qdrant health check failed: %s", e)
        services.append(ServiceStatus(
            name="qdrant", status="error", detail=str(e)
        ))
        overall_healthy = False

    # Full metrics report for the dashboard (Day 42)
    report = None
    try:
        monitor = HealthMonitor(driver)
        report = monitor.full_health_report()
    except Exception as e:
        logger.error("Health monitor full report failed: %s", e)
        # Don't flip overall_healthy to False here — connectivity is already
        # verified above; this is richer detail, not a core health check.


    return HealthResponse(
        status="healthy" if overall_healthy else "degraded",
        services=services,
        counts=counts,
        report=report,
    )